### DISCUSSION TITLE
Docker Container at setup quite big

### DISCUSSION THREAD

**Discussion by gsg-git** [mediagis/nominatim-docker#477] [Q&A] [answered]
Hello!

I use for the Server a 2TB drive and had uptil now no size problems.
I aleredy uploaded the World and at last ran the index script after the plz one, at friday and see today, that the script got interrupted, because the drive is full (right now 1.6 from 1.8 TB) and i just want to know if thi is a normal behaviour?

Edit:
I can see the PostgreSQL in the conainer is the one that uses all the space up : 1,5TB, but i still have no idea why:
1.5T    /var/lib/postgresql/14/main/base/16386

**Comment by mtmail** [comment]
Not normal, 1TB should already be enough. Do you have any other old databases on the drive maybe? Check „docker container ls“ and „docker image ls“ if there’s old docker data you might no longer need

**Reply by gsg-git** [reply]
I ckecked in the Postgres Database and the Nominatim Databade ist this big - now i somehow have to find out which Table uses all the space

**Reply by gsg-git** [reply]
The server is fresh and i deleted everything in Docker after each failure and the size i mentioned of the Postgres folder is in the Nominatim Container itself.
 In PostgreSQL:
nominatim=#  \l+
                                                                    List of databases
   Name    |   Owner   | Encoding |   Collate   |    Ctype    |   Access privileges   |  Size   | Tablespace |                Description
-----------+-----------+----------+-------------+-------------+-----------------------+---------+------------+--------------------------------------------
 nominatim | nominatim | UTF8     | en_US.UTF-8 | en_US.UTF-8 |                       | 1508 GB | pg_default |
 postgres  | postgres  | UTF8     | en_US.UTF-8 | en_US.UTF-8 |                       | 8569 kB | pg_default | default administrative connection database
 template0 | postgres  | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +| 8401 kB | pg_default | unmodifiable empty database
           |           |          |             |             | postgres=CTc/postgres |         |            |
 template1 | postgres  | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +| 8553 kB | pg_default | default template for new databases
           |           |          |             |             | postgres=CTc/postgres |         |            |

I looked the sizes up and there are around 600 and those are thebiggest in Order:
          table_name         | pg_size_pretty | pg_relation_size
---------------------------+----------------+------------------
 planet_osm_nodes          | 356 GB         |     382769725440
 planet_osm_ways           | 207 GB         |     222491418624
 placex                    | 166 GB         |     178432720896
 place                     | 98 GB          |     104881102848
 search_name               | 62 GB          |      66674573312
 place_addressline         | 34 GB          |      36098949120
 word                      | 6257 MB        |       6560817152
 planet_osm_rels           | 4793 MB        |       5025423360
 search_name_2             | 4328 MB        |       4538048512
 location_road_2           | 3572 MB        |       3745325056
 search_name_4             | 1270 MB        |       1331691520
 search_name_3             | 1054 MB        |       1104699392
 location_road_4           | 865 MB         |        906592256
 location_road_3           | 743 MB         |        779304960
 search_name_121           | 742 MB         |        778395648
 search_name_31            | 607 MB         |        636149760
 search_name_28            | 599 MB         |        627621888
 search_name_1             | 577 MB         |        605028352
 search_name_117           | 566 MB         |        593698816
 location_road_121         | 537 MB         |        562839552
 search_name_135           | 509 MB         |        533372928
 location_property_osmline | 464 MB         |        486203392
 search_name_244           | 455 MB         |        477175808
 location_road_28          | 453 MB         |        475308032
 search_name_139           | 412 MB         |        432193536
 location_road_31          | 401 MB         |        420937728
 location_road_1           | 399 MB         |        418291712
 search_name_166           | 388 MB         |        407240704
 location_road_244         | 358 MB         |        375316480
 location_road_117         | 358 MB         |        374939648
 location_area_large_4     | 351 MB         |        367869952
 location_road_139         | 316 MB         |        331202560
 location_road_135         | 310 MB         |        324558848
 location_road_166         | 290 MB         |        303734784
 search_name_168           | 289 MB         |        303161344
 search_name_11            | 268 MB         |        280690688
 location_postcode         | 262 MB         |        274481152
 search_name_63            | 249 MB         |        261537792
 location_area_large_135   | 235 MB         |        246571008


but i cant tell if any of this is correct.

I rather wait for a hint, because  creating the index took over 17 hours and i rather wait bevor the script stops again because of the full drive again.
My steps where:
1. After the first failed build i deleted every part of the old conainer and image and datastore
2. I ran the :
docker run -it \
  -e PBF_URL=http://download.geofabrik.de/africa-latest.osm.pbf \
  -e REPLICATION_URL=https://download.geofabrik.de/africa-updates/\
  -p 8080:8080 \
  --name nominatim \
  mediagis/nominatim:4.2
3. Tested the build successfully with an HTTP request
4. I connected to the conatiner, created in /nominatim a folder for the map downloads and downladed the rest of the countrys osm files
5. switched to the nominatim user and imported the other osm files: nominatim add-data --file europe-latest.osm.pbf
6. Started from /nominatim as nominatim those two: nominatim refresh --postcodes 
    6.1 nominatim index -j 8 
and the last step got canceled because of the full drive.
For adding further data i too the Advanced installation guide:
https://nominatim.org/release-docs/latest/admin/Advanced-Installations/


Thank you in advance very much for your input!

**Reply by mtmail** [reply, marked as answer]
> planet_osm_nodes | 356 GB

https://nominatim.org/release-docs/latest/admin/Import/ and the Docker readme file say for continents and whole planet you should use the flatnode file. That's likely the exact 250-300GB. With the flatnode file the planet_osm_node tables doesn't get created and import is also a bit faster

>  because creating the index took over 17 hours

That's not a lot. Depending on hardware people wait 2-7 days, I've heard of 14 days even.

**Reply by gsg-git** [reply]
@mtmail 
Tank you!
I thought the flatnode would be created in the docker version, but there is just the path settings.
But can i now reverse my mistake?
And how do i set the flatnode file? in the guide i think its just a text file named flatnode.file and i have to drop it into /nominatim/flatnode after the first Setup of the nominatim Docker setup?
What if i want to import the World in the first run how could i set the flatnode then?

Like this?

docker run -it \
  -e PBF_URL=https://ftp5.gwdg.de/pub/misc/openstreetmap/planet.openstreetmap.org/pbf/planet-latest.osm.pbf \
  -e REPLICATION_URL=https://ftp5.gwdg.de/pub/misc/openstreetmap/planet.openstreetmap.org/replication/day/ \
  -e NOMINATIM_FLATNODE_FILE=/nominatim/flatnode/flatnode.file \
  -p 8080:8080 \
  --name nominatim \
  mediagis/nominatim:3.7 

And whats questioning me in the end is:
I didnt use a flatnode file in the first vanilla installation, but the database wasnt this big ater i hat imported the world.
The only diffrence was i ran the PLZ and inxexing after each import.
Maybe there are some data which where cleaned at the end of the indexing process and now with one  indexing for all the temp. stuff got too big for the drive.

**Reply by edurenye** [reply]
> > planet_osm_nodes | 356 GB
> 
> https://nominatim.org/release-docs/latest/admin/Import/ and the Docker readme file say for continents and whole planet you should use the flatnode file. That's likely the exact 250-300GB. With the flatnode file the planet_osm_node tables doesn't get created and import is also a bit faster
> 
> > because creating the index took over 17 hours
> 
> That's not a lot. Depending on hardware people wait 2-7 days, I've heard of 14 days even.

I had a weird experience with flatnode, the DB was 10GB big, just got a few regions of Spain and when I used flatnode, the volume grow to more than 100GB and the disk ran out of space. Feels more like a bug, but I'm not sure, I'm using version 5.1.

**Reply by mtmail** [reply]
The flatnode file is always 100GB regardless which region gets imported. https://osm2pgsql.org/doc/manual.html#flat-node-store

**Reply by edurenye** [reply]
Oh, good to know, thank you!
